### PR TITLE
Add SessionPersistence API as experimental

### PR DESCRIPTION
**Pull request by gcs278**
From GEP-1619, add BackendLBPolicy with SessionPersistence fields and add the SessionPersistence fields on HTTPRouteRule and GRPCRouteRule as experimental.

/kind feature

**What this PR does / why we need it**:
Makes SessionPersistence API experimental.

**Which issue(s) this PR fixes**:
Updates #713

**Does this PR introduce a user-facing change?**:

```release-note
BackendLBPolicy has been added to enable session persistence configuration for backends.
A SessionPersistence field has been added to HTTPRoute and GRPCRoute to enable session persistence for routes.
```

### CODE REVIEW

File: apis/v1/shared_types.go
Comment (gcs278) [inline apis/v1/shared_types.go:827]: Should this requirement of having `SessionPersistence.AbsoluteTimeout` when `SessionPersistence.LifetimeType = Permanent` be enforced by CEL?

Or should we just allow implementations to decide a valid default?

Comment (robscott) [reply apis/v1/shared_types.go:827]: Good question, I think we should definitely aim to enforce this with CEL. Happy to help figure out what that looks like if it ends up being tricky.

File: apis/v1/shared_types.go
Comment (robscott) [inline apis/v1/shared_types.go:758]: This seems like an awfully long length, what kind of session name would be this long?

File: apis/v1alpha2/backendlbpolicy_types.go
Comment (robscott) [inline apis/v1alpha2/backendlbpolicy_types.go:59]: I know you're already on this, but just in case we forget, when #2966 merges, we'll need to update this to be both a list and a local reference.

**Review by robscott — COMMENTED** [review]
Thanks @gcs278! Just took a quick look at this but so far mostly LGTM.

File: apis/v1/shared_types.go
Comment (gcs278) [reply apis/v1/shared_types.go:758]: Good question...

I believe I got the max length from https://www.rfc-editor.org/rfc/rfc6265:
>   Practical user agent implementations have limits on the number and
   size of cookies that they can store.  General-use user agents SHOULD
   provide each of the following minimum capabilities:
>
>    o  At least 4096 bytes per cookie (as measured by the sum of the
      length of the cookie's name, value, and attributes).

But this is the _minimum capabilities_, so I suppose that's the worst case (smallest supported cookie size)...of the worst case (the entire cookie size is the name...).

So yea 4096 is a somewhat arbitrary. Is it best practice to limit string lengths? Or is it okay to leave unspecified?

File: apis/v1/shared_types.go
Comment (gcs278) [reply apis/v1/shared_types.go:828]: Added it.

I tested the CEL by hand in my own cluster with the CRD. Are there CRD integration tests for validating CEL that I should add a test case to? I looked around and couldn't find anything.

File: apis/v1/shared_types.go
Comment (robscott) [reply apis/v1/shared_types.go:759]: >Is it best practice to limit string lengths?

Yep, it's been a strong requirement to avoid any kind of unbounded values in Gateway API.

> as measured by the sum of the length of the cookie's name, value, and attributes

In the case of cookie-based persistence this would just be the cookie name right? (Not the value or attributes)

File: apis/applyconfiguration/apis/v1/sessionpersistence.go
Comment (candita) [inline apis/applyconfiguration/apis/v1/sessionpersistence.go:1]: As an experimental feature, does this belong in v1?

Comment (robscott) [reply apis/applyconfiguration/apis/v1/sessionpersistence.go:1]: Discussed in meeting, but since this is an experimental field and not an entire resource, we have to add it to any API versions that exist for that resource, in this case that includes v1. Importantly we add the `<gateway:experimental>` tags around this field to ensure it's only accessible if you install the experimental version of this CRD. This is conceptually similar to upstream Kubernetes where new fields added to the v1 Service API are initially guarded by alpha feature gates (but if that feature gate is enabled, the field is accessible as part of the v1 Service API).

File: apis/v1/shared_types.go
Comment (gcs278) [reply apis/v1/shared_types.go:759]: > In the case of cookie-based persistence this would just be the cookie name right? (Not the value or attributes)

Correct. `SessionName` is just the cookie name. I was just trying to find established limits, and the only one I came across was this 4096 bytes value for the *whole* cookie.

Since I can't find any enforced value, should we start with something reasonable? Maybe 128 or 256?

**Comment by gcs278** [comment]
/cc @kflynn

@robscott Updated to include `LocalPolicyTargetReference` as provided by https://github.com/kubernetes-sigs/gateway-api/pull/2966

File: apis/v1/shared_types.go
Comment (robscott) [reply apis/v1/shared_types.go:828]: Yep, sorry I missed this, these tests are part of https://github.com/kubernetes-sigs/gateway-api/tree/main/pkg/test/cel. Here's where we run them in presubmits: https://github.com/kubernetes-sigs/gateway-api/blob/a101a237332dbb9216cf5d8b85acc238ba56d4cf/hack/verify-crds-kind.sh#L60

File: apis/v1alpha2/backendlbpolicy_types.go
Comment (robscott) [inline apis/v1alpha2/backendlbpolicy_types.go:59]:
```suggestion
	//+listType=atomic
	// +kubebuilder:validation:MinItems=1
	// +kubebuilder:validation:MaxItems=16
	TargetRefs []LocalPolicyTargetReference `json:"targetRef"`
```

I think this should be a list of object references, and I think this list should be "atomic" as I'd expect to exclusively have one owner, but interested in what others think here.

File: apis/v1/shared_types.go
Comment (robscott) [reply apis/v1/shared_types.go:759]: Let's start with 128 and increase if/when necessary. It's hard for me to think of a use case that would require a longer session name, but I could be missing something.

**Review by robscott — COMMENTED** [review]
Thanks @gcs278!

File: apis/v1alpha2/backendlbpolicy_types.go
Comment (robscott) [reply apis/v1alpha2/backendlbpolicy_types.go:59]: cc @dprotaso for advice on list type

Comment (dprotaso) [reply apis/v1alpha2/backendlbpolicy_types.go:59]: listType=map with keys [group, kind, name] would make sense here. Then different parts of the list could be 'owned' by different people

Comment (robscott) [reply apis/v1alpha2/backendlbpolicy_types.go:59]: Sounds good, that likely is far more future proof. Somehow I'd convinced myself that these resources would always have a single owner, but there will likely be at least some edge cases where that's not true. I've updated the original suggestion to reflect your advice @dprotaso.

File: apis/v1/shared_types.go
Comment (gcs278) [reply apis/v1/shared_types.go:828]: Ah should I looked a little harder...added a CEL test.

File: apis/v1alpha2/backendlbpolicy_types.go
Comment (gcs278) [reply apis/v1alpha2/backendlbpolicy_types.go]: Ack. Updated to be a list.

**Comment by robscott** [comment]
Thanks @gcs278! Will defer to someone else for final LGTM.

/approve

File: apis/v1/shared_types.go
Comment (shaneutt) [inline apis/v1/shared_types.go:748]:
```suggestion
// SessionPersistence defines the desired state of SessionPersistence.
```

File: apis/v1/shared_types.go
Comment (shaneutt) [inline apis/v1/shared_types.go:784]: :+1:

File: apis/v1/shared_types.go
Comment (shaneutt) [inline apis/v1/shared_types.go:786]: :+1:

File: apis/v1/shared_types.go
Comment (shaneutt) [inline apis/v1/shared_types.go:795]: :+1:

**Review by shaneutt — APPROVED** [review]
Seems like a great huge step to getting something out there that implementations can try to implement and start giving us some feedback :+1: 

/approve

@gcs278 out of curiosity do you have an implementation that you tested this out with? If you do, would you be opposed to putting some time on our meeting agenda to do a quick demo for us? Would be really cool!

(also in general if you've had a chance to implement this locally and sort of try it out, that could be very valuable)

File: apis/v1/shared_types.go
Comment (gcs278) [reply apis/v1/shared_types.go:748]: Done.

**Comment by gcs278** [comment]
> Seems like a great huge step to getting something out there that implementations can try to implement and start giving us some feedback 👍
> 
> /approve
> 
> @gcs278 out of curiosity do you have an implementation that you tested this out with? If you do, would you be opposed to putting some time on our meeting agenda to do a quick demo for us? Would be really cool!

I do not. I would love for an implementation to give this a shot. I will advertise for it in our community meeting. 

> (also in general if you've had a chance to implement this locally and sort of try it out, that could be very valuable)

+1 - that's one of the reasons I brought up Blitx as a reference implementation. I wish I had a "proving ground" for testing the API. I suppose it can be a real implementation. One of my concerns with the experimental API, is that it's mostly theoretical. I'm open to suggestions.

**Comment by robscott** [comment]
Thanks @gcs278!

/lgtm
/hold cancel

**Pull request merged**
